lib/std/private/strimpl.nim

```diff
@@ -0,0 +1,41 @@
+func toLowerAscii*(c: char): char {.inline.} =
+  if c in {'A'..'Z'}:
+    result = chr(ord(c) + (ord('a') - ord('A')))
+  else:
+    result = c
+
+template cmpIgnoreStyleImpl*(a, b: typed, caseSensitive: static bool = false) =
+  # a, b are string or cstring
+  when caseSensitive:
+    if a[0] != b[0]: return 1
+  var i = 0
+  var j = 0
+  while true:
+    while i < a.len and a[i] == '_': inc i
+    while j < b.len and b[j] == '_': inc j
+    let aa = if i < a.len: toLowerAscii(a[i]) else: '\0'
+    let bb = if j < b.len: toLowerAscii(b[j]) else: '\0'
+    result = ord(aa) - ord(bb)
+    if result != 0: return result
+    # the characters are identical:
+    if i >= a.len:
+      # both cursors at the end:
+      if j >= b.len: return 0
+      # not yet at the end of 'b':
+      return -1
+    elif j >= b.len:
+      return 1
+    inc i
+    inc j
+
+template cmpIgnoreCaseImpl*(a, b: typed, caseSensitive: static bool = false) =
+  # a, b are string or cstring
+  when caseSensitive:
+    if a[0] != b[0]: return 1
+  var i = 0
+  var m = min(a.len, b.len)
+  while i < m:
+    result = ord(toLowerAscii(a[i])) - ord(toLowerAscii(b[i]))
+    if result != 0: return
+    inc(i)
+  result = a.len - b.len
```

Comment:
```suggestion
template cmpIgnoreStyleImpl*(a, b: typed, caseSensitive: static bool = false) =
  # a, b are string or cstring
  var aLen = a.len
  var bLen = b.len
  when caseSensitive:
    if a[0] != b[0]: return 1
  var i = 0
  var j = 0
  while true:
    while i < aLen and a[i] == '_': inc i
    while j < bLen and b[j] == '_': inc j
    let aa = if i < aLen: toLowerAscii(a[i]) else: '\0'
    let bb = if j < bLen: toLowerAscii(b[j]) else: '\0'
    result = ord(aa) - ord(bb)
    if result != 0: return result
    # the characters are identical:
    if i >= aLen:
      # both cursors at the end:
      if j >= bLen: return 0
      # not yet at the end of 'b':
      return -1
    elif j >= bLen:
      return 1
    inc i
    inc j

template cmpIgnoreCaseImpl*(a, b: typed, caseSensitive: static bool = false) =
  # a, b are string or cstring
  var aLen = a.len
  var bLen = b.len
  when caseSensitive:
    if a[0] != b[0]: return 1
  var i = 0
  var m = min(aLen, bLen)
  while i < m:
    result = ord(toLowerAscii(a[i])) - ord(toLowerAscii(b[i]))
    if result != 0: return
    inc(i)
  result = aLen - bLen
```

Reply:
Because `len` is O(n) for cstrings.

Reply:
Sure. But cstring is only used in JS backend(at least now and hard to be used in other places).
cstring in JS backend is O(1).

Reply:
I'm not sure cstring.len in VM.

Reply:
Ah ok, still can't hurt.

Reply:
Yes, I agree.

Reply:
refs: https://github.com/nim-lang/Nim/issues/16405